arctic/store/metadata_store.py

```diff
@@ -0,0 +1,217 @@
+from datetime import datetime as dt
+import logging
+
+import pandas as pd
+
+import bson
+import pymongo
+from ..decorators import mongo_retry
+from ..exceptions import NoDataFoundException
+
+from .bson_store import BSONStore
+
+logger = logging.getLogger(__name__)
+
+METADATA_STORE_TYPE = 'MetadataStore'
+
+class MetadataStore(BSONStore):
+    """
+    Metadata Store. This stores metadata with timestamps to allow temporal queries.
+    """
+
+    @classmethod
+    def initialize_library(cls, arctic_lib, hashed=True, **kwargs):
+        MetadataStore(arctic_lib)._ensure_index()
+        BSONStore.initialize_library(arctic_lib, hashed, **kwargs)
+
+    @mongo_retry
+    def _ensure_index(self):
+        self.create_index([('symbol', pymongo.ASCENDING), ('start_time', pymongo.DESCENDING)],
+                          unique=True, background=True)
+
+    def __init__(self, arctic_lib):
+        self._arctic_lib = arctic_lib
+        self._collection = self._arctic_lib.get_top_level_collection()
+
+    @mongo_retry
+    def find_one(self, *args, **kwargs):
+        """
+        See http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.find_one
+        """
+        return self._collection.find_one(*args, **kwargs)
+
+    @mongo_retry
+    def find_one_and_update(self, filter, update, **kwargs):
+        """
+        See http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.find_one_and_update
+        """
+        self._arctic_lib.check_quota()
+        return self._collection.find_one_and_update(filter, update, **kwargs)
+
+    def find_one_and_delete(self, filter, **kwargs):
+        """
+        See http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.find_one_and_delete
+        """
+        return self._collection.find_one_and_delete(filter, **kwargs)
+
+    def list_symbols(self):
+        return self.distinct('symbol')
+
+    def has_symbol(self, symbol):
+        return self.find_one({'symbol': symbol}) is not None
+
+    def read_metadata(self, symbol, history=False):
```

Comment:
I would've implement this as a dataframe store, and return the last entry if `history=False`.
But this works as well :)